proc/threads.c

```diff
@@ -1366,47 +1557,81 @@ int proc_lockSet2(lock_t *l1, lock_t *l2)
 {
 	int err;
 
-	if ((err = proc_lockSet(l1)) < 0)
+	if ((err = proc_lockSet(l1)) < 0) {
 		return err;
+	}
 
 	while (proc_lockTry(l2) < 0) {
 		proc_lockClear(l1);
-		if ((err = proc_lockSet(l2)) < 0)
+		if ((err = proc_lockSet(l2)) < 0) {
 			return err;
+		}
 		swap(l1, l2);
 	}
+
 	return EOK;
 }
 
 
 int proc_lockWait(thread_t **queue, lock_t *lock, time_t timeout)
 {
-	int err;
 	spinlock_ctx_t sc;
+	int err;
 
 	hal_spinlockSet(&lock->spinlock, &sc);
+
 	_proc_lockClear(lock);
-	if ((err = proc_threadWaitEx(queue, &lock->spinlock, timeout, 1, &sc)) != -EINTR)
+	if ((err = proc_threadWaitEx(queue, &lock->spinlock, timeout, 1, &sc)) != -EINTR) {
 		_proc_lockSet(lock, 0, &sc);
+	}
+
 	hal_spinlockClear(&lock->spinlock, &sc);
+
 	return err;
 }
 
 
-int proc_lockInit(lock_t *lock)
+static int _proc_lockDone(lock_t *lock)
 {
-	lock->owner = NULL;
-	lock->priority = 0;
-	lock->queue = NULL;
-	lock->v = 1;
-	hal_spinlockCreate(&lock->spinlock, "lock.spinlock");
+	spinlock_ctx_t sc;
+	thread_t *owner;
+	int refs;
+
+	if ((owner = lock->owner) != NULL) {
+		hal_spinlockSet(&owner->spinlock, &sc);
+
+		LIST_REMOVE(&owner->locks, lock);
+		refs = --owner->refs;
+
+		hal_spinlockClear(&owner->spinlock, &sc);
+
+		if (refs == 0) {
+			thread_destroy(owner);
+		}
```

Comment:
We also need to wake up everyone waiting to take the lock and cause them to exit with error

Reply:
Done but I'm not sure if my solution really works - I'm going to write a specific test for this scenario.

Reply:
It turned out that `proc_lockDone()` is always called with empty `lock->queue` and there are no waiting threads to wake up.